api/catalog/api/controllers/elasticsearch/search.py

```diff
@@ -0,0 +1,198 @@
+    """
+    If there are any unmatched quotes in the query supplied by the user, ignore
+    them by escaping.
+
+    :param query_string: the string in which to escape unbalanced quotes
+    :return: the given string, if the quotes are balanced, the escaped string otherwise
+    """
+
+    num_quotes = query_string.count('"')
+    if num_quotes % 2 == 1:
+        return query_string.replace('"', '\\"')
+    else:
+        return query_string
+
+
+def _apply_filter(
+    s: Search,
+    query_serializer: MediaSearchRequestSerializer,
+    basis: Union[str, tuple[str, str]],
+    behaviour: Literal["filter", "exclude"] = "filter",
+) -> Search:
+    """
+    Parse and apply a filter from the search parameters serializer. The
+    parameter key is assumed to have the same name as the corresponding
+    Elasticsearch property. Each parameter value is assumed to be a comma
+    separated list encoded as a string.
+
+    :param s: the search query to issue to Elasticsearch
+    :param query_serializer: the ``MediaSearchRequestSerializer`` object with the query
+    :param basis: the name of the field in the serializer and Elasticsearch
+    :param behaviour: whether to accept (``filter``) or reject (``exclude``) the hit
+    :return: the modified search query
+    """
+
+    search_params = query_serializer.data
+    if isinstance(basis, tuple):
+        serializer_field, es_field = basis
+    else:
+        serializer_field = es_field = basis
+    if serializer_field in search_params:
+        filters = []
+        for arg in search_params[serializer_field].split(","):
+            filters.append(Q("term", **{es_field: arg}))
+        method = getattr(s, behaviour)  # can be ``s.filter`` or ``s.exclude``
+        return method("bool", should=filters)
+    else:
+        return s
+
+
+def perform_search(
+    query_serializer: MediaSearchRequestSerializer,
+    index: Literal["image", "audio"],
+    ip: int,
+) -> Tuple[List[Hit], int, int]:
+    """
+    Perform a ranked, paginated search based on the query and filters given in the
+    search request.
+
+    :param query_serializer: the ``MediaSearchRequestSerializer`` object with the query
+    :param index: The Elasticsearch index to search (e.g. 'image')
+    :param ip: the users' hashed IP to consistently route to the same ES shard
+    :return: the list of search results with the page and result count
+    """
+
+    s = Search(using="default", index=index)
+    search_params = query_serializer.data
+
+    rules: dict[Literal["filter", "exclude"], list[Union[str, tuple[str, str]]]] = {
+        "filter": [
+            "extension",
+            "category",
+            ("categories", "category"),
+            "aspect_ratio",
+            "size",
+            "length",
+            "source",
+            ("license", "license.keyword"),
+            ("license_type", "license.keyword"),
+        ],
+        "exclude": [
+            ("excluded_source", "source"),
+        ],
```

Comment:
This seems like something we'll have to edit when adding more media types or changing available columns - do you think it might be best in a `constants` module or somewhere alongside other lists of available columns? I worry it'll be overlooked when adding something.

Reply:
I'll make a separate issue to improve this filtering setup, this PR is blocking others.